vertical-pod-autoscaler/initializer/initializer.go

```diff
@@ -0,0 +1,347 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	watch "k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/admissionregistration/v1alpha1"
+
+	"time"
+
+	kubeclient "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	admissionregistrationv1alpha1 "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/admissionregistration/v1alpha1"
+
+	"github.com/golang/glog"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/apimock"
+	recommender "k8s.io/autoscaler/vertical-pod-autoscaler/recommender_mock"
+)
+
+const (
+	vpaInitializerConfigName = "vpa-initializer-config.k8s.io"
+	vpaInitializerName       = "vpa-initializer.k8s.io"
+)
+
+// Initializer initializes all pods matched by an existing vpa object
+// by assigning them recommended resource requests.
+type Initializer interface {
+	Run(stopCh <-chan struct{})
+}
+
+type initializer struct {
+	client          kubeclient.Interface
+	vpaLister       apimock.VerticalPodAutoscalerLister
+	podListerSynced cache.InformerSynced
+	informer        cache.SharedInformer
+	recommender     recommender.CachingRecommender
+	registerer      admissionregistrationv1alpha1.InitializerConfigurationInterface
+}
+
+func (i *initializer) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting VPA initializer.")
+	defer glog.Infof("Shutting down VPA initializer.")
+
+	i.register()
+	defer i.unregister()
+	i.informer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, i.podListerSynced) {
+		glog.Error("Cache sync failed.")
+		return
+	}
+}
+
+func (i *initializer) register() error {
+	configuration := newConfiguration()
+	config, err := i.registerer.Create(configuration)
+	if err != nil {
+		//TODO: If exists, return nil
+		glog.Infof("Error registering VPA Initializer Config: %v", err)
+		return err
+	}
+	glog.V(4).Infof("Registered config: %+v", config)
+	return nil
+}
+
+func (i *initializer) unregister() {
+	i.registerer.Delete(vpaInitializerConfigName, &metav1.DeleteOptions{})
+}
+
+func (i *initializer) updateResourceRequests(obj interface{}) {
+	pod := obj.(*v1.Pod)
+
+	if !shouldInitialize(pod) {
+		glog.V(4).Infof("Not updating pod: %v", pod.Name)
+		return
+	}
+	initializedPod, err := i.getInitializedPod(pod)
+
+	if err == nil {
+		// TODO(bskiba): Right now this update always fails if we try to change resource requests
+		// as there are checks in place that prevent from changing requests of a created pod.
+		result, err := i.client.CoreV1().Pods(pod.Namespace).Update(initializedPod)
+		if err == nil {
+			glog.V(3).Infof("Successfully updated pod %+v", result)
+			return
+		}
+	}
+	failedPod, err := getFailedPod(pod)
+	if err == nil {
+		i.client.CoreV1().Pods(pod.Namespace).Update(failedPod)
+	} else {
+		log.Fatalf("Unable to mark failed initialization for pod %v: %v", pod.Name, err)
```

Comment:
why handle error==nil case first? usually we handle error case first when it is a fatal issue.